src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,151 @@
+            TypeSyntax,
+            VariableDeclaratorSyntax>
+    {
+        protected override string GetTitle()
+            => CSharpFeaturesResources.Convert_for_to_foreach;
+
+        protected override SyntaxList<StatementSyntax> GetBodyStatements(ForStatementSyntax forStatement)
+            => forStatement.Statement is BlockSyntax block
+                ? block.Statements
+                : SyntaxFactory.SingletonList(forStatement.Statement);
+
+        protected override bool TryGetForStatementComponents(
+            ForStatementSyntax forStatement,
+            out SyntaxToken iterationVariable, out ExpressionSyntax initializer,
+            out MemberAccessExpressionSyntax memberAccess, 
+            out ExpressionSyntax stepValueExpressionOpt,
+            CancellationToken cancellationToken)
+        {
+            // Look for very specific forms.  Basically, only minor variations around:
+            // for (var i = 0; i < expr.Lenth; i++)
+
+            if (forStatement.Declaration != null &&
+                forStatement.Condition.IsKind(SyntaxKind.LessThanExpression) &&
+                forStatement.Incrementors.Count == 1)
+            {
+                var declaration = forStatement.Declaration;
+                if (declaration.Variables.Count == 1)
+                {
+                    var declarator = declaration.Variables[0];
+                    if (declarator.Initializer != null)
+                    {
+                        iterationVariable = declarator.Identifier;
+                        initializer = declarator.Initializer.Value;
+
+                        var binaryExpression = (BinaryExpressionSyntax)forStatement.Condition;
+
+                        // Look for:  i < expr.Length
+                        if (binaryExpression.Left is IdentifierNameSyntax identifierName &&
+                            identifierName.Identifier.ValueText == iterationVariable.ValueText &&
+                            binaryExpression.Right is MemberAccessExpressionSyntax)
+                        {
+                            memberAccess = (MemberAccessExpressionSyntax)binaryExpression.Right;
+
+                            var incrementor = forStatement.Incrementors[0];
+                            return TryGetStepValue(
+                                iterationVariable, incrementor, out stepValueExpressionOpt, cancellationToken);
+                        }
+                    }
+                }
+            }
+
+            iterationVariable = default;
+            memberAccess = default;
+            initializer = default;
+            stepValueExpressionOpt = default;
+            return false;
+        }
+
+        private static bool TryGetStepValue(
+            SyntaxToken iterationVariable, ExpressionSyntax incrementor,
+            out ExpressionSyntax stepValue, CancellationToken cancellationToken)
+        {
+            // support
+            //  x++
+            //  ++x
+            //  x += constant_1
+
```

Comment:
What about 
i = i + 1
i = i++

Reply:
> ```i = i++```

I have never see that in C# code, and i would be 100% ok not supporting that case.

> ```i = i + 1```

That's at least somewhat reasonable.  But, for now, is not something detected.  i don't think ta'ts common enough to warrant looking for.

Reply:
> i = i++

that should be a compiler error :smiley:

Reply:
> i = i + 1

Feels natural if you're coming from a Pascal or VB(A) background.

I think if you support it, you could also support the weird i = i++ as you could take the right hand side of the assignment and let it go through your existing logic.

But I also agree that this isn't that important.